apps/prairielearn/src/middlewares/selectAndAuthzInstanceQuestion.sql

```diff
@@ -1,5 +1,45 @@
 -- BLOCK select_and_auth
 WITH
+  group_role_ids AS (
+    SELECT
+      group_role_id
+    FROM
+      group_user_roles AS gur
+      JOIN groups AS g ON gur.group_id = g.id
+      JOIN assessment_instances AS ai ON g.id = ai.group_id
+      JOIN instance_questions AS iq ON (iq.assessment_instance_id = ai.id)
+    WHERE
+      gur.user_id = $user_id
+      AND iq.id = $instance_question_id
+  ),
+  visibility_info AS (
+    SELECT
+      iq.id AS instance_question_id,
+      bool_or(aqrp.can_view) AS can_user_view,
+      bool_or(aqrp.can_submit) AS can_user_submit
+    FROM
+      assessment_instances AS ai
+      JOIN instance_questions AS iq ON (iq.assessment_instance_id = ai.id)
+      JOIN assessment_questions AS aq ON (aq.id = iq.assessment_question_id)
+      JOIN assessment_question_role_permissions AS aqrp ON (aqrp.assessment_question_id = aq.id)
+    WHERE
+      ai.id IN (
+        SELECT
+          assessment_instance_id
+        FROM
+          instance_questions
+        WHERE
+          id = $instance_question_id
+      )
```

Comment:
I'm not sure that replacement would work. For the `visibility_info` CTE, we want to get the user's `can_view` and `can_submit` permissions for _all_ questions in the current assessment so that we can link/disable the "Next/Previous Question" buttons accordingly. Since we don't have the current assessment instance's ID as a parameter, this block looks it up from the `instance_question_id`.

Reordering the `FROM` clause to put `iq` first makes sense - will change.

Reply:
Ahh, I see that now - it wasn't clear to me that this was returning information about all questions. Yeah, my suggestion won't work in that case, sorry!

To check my understanding: we join with `visibility_info` below so that we can filter out un-viewable questions when computing the previous/next questions?

Reply:
Yup!

Reply:
Gotcha - that makes sense!